validator.go

```diff
@@ -117,13 +117,22 @@ func ValidateAll(xmlReader io.Reader) []error {
 
 // bufio implements a ByteReader but we explicitly don't want any buffering
 type byteReader struct {
-	r io.Reader
+	r   io.Reader
 }
 
 func (r *byteReader) ReadByte() (byte, error) {
-	p := make([]byte, 1)
-	_, err := r.r.Read(p)
-	return p[0], err
+	var p [1]byte
+	n, err := r.r.Read(p[:])
+
+	// the doc for the io.ReadeByte interface states :
+	//   If ReadByte returns an error, no input byte was consumed, and the returned byte value is undefined
+	// so : if a byte is actually extracted from the reader, and we want to return it, we mustn't return the error
+	if n > 0 {
+		// return the byte read from the reader
+		return p[0], nil
+	}
+
+	return 0, err
```

Comment:
This is slightly tricky. I would say our handling of io.Reader is actually incorrect here. The docs for that [say](https://golang.org/pkg/io/#Reader)

> When Read encounters an error or end-of-file condition after successfully reading n > 0 bytes, it returns the number of bytes read. It may return the (non-nil) error from the same call or return the error (and n == 0) from a subsequent call. An instance of this general case is that a Reader returning a non-zero number of bytes at the end of the input stream may return either err == EOF or err == nil. The next Read should return 0, EOF.

> Callers should always process the n > 0 bytes returned before considering the error err. 

Basically, when the last byte is read, it may return io.EOF in that call, or in a subsequent call. And as recommended, we should always process the returned bytes, before returning the error.

With that in mind, I would suggest to actually add the check for `n > 0 && io.EOF`. This matches the expectation of io.Reader more closely, essentially ensuring that if an EOF is returned in the case where `n > 0` bytes are returned, we wipe that error to remain consistent with the io.ByteReader implementation.

Everything else may remain the same. Here is my suggested change:

```go
var p [1]byte
n, err := r.r.Read(p[:])
// When the last byte is read, it may return io.EOF in that call, or in a subsequent call.
// To comply with a io.ByteReader implementation, we gobble up the io.EOF.
if n > 0 && err == io.EOF {
	return p[0], nil
}
return p[0], err
```

Reply:
Hi @agnivade,

I understand the idea to not hide the errors, your suggestion definitely goes towards that. I'll restate that I am fine with either of the approaches :

1. somehow return errors to the caller,
2. or completely discard errors that come together with 1 byte extracted from the Reader.

I would discuss the "hide EOF and don't hide other errors" point though :  
if this implementation of `io.ByteReader` was a generic implementation intended for all use cases, I wouldn't treat EOF differently from other errors. What if the last call to `.Readbyte()` returned `'>', {"i/o timeout"}` ? or `'>', {some syscall.Errno}` ? or `'>', {some custom error}` ? These would also trigger the "unexpected EOF" error, when we actually have that final `'>'`.

A reference implementation would be `bufio.Reader`, which caches and returns any error (not just EOF).  
This is also the approach I followed in the initial version of this PR.

I would also highlight that this specific implementation is not generic : it is a private implementation, used in the specific case of `Validate()`.

For example : `Validate()` consumes an xml payload from the Reader without returning it, so it implicitly relies on the fact that the caller has a way to retrieve that payload afterwards.  
Your examples suggest to feed a `bytes.NewReader([]byte)`, we hit our bug because we thought we thought we were clever and chose to wrap our existing Reader in a `io.TeeReader()` (much like what is done in the `Validate()` implementation), and you probably have other use cases in your own code base.

And in that context : I think that approach 2. as @jupenur suggested is also valid.

I'm mentioning these two points as food for thought ; with that in mind, what would be your advice on this behavior ?

Reply:
Fair enough, gave some more thought into this. Let's go with approach 2. Sorry for the to-and-fro there. Great discussion. :+1: 

Should be good to go once we fix https://github.com/mattermost/xml-roundtrip-validator/pull/7#discussion_r545577786 and any gofmt issues.

Thanks @LeGEC for your time on this.